File: RunLengthEncode.py
class RunLengthEncode:
    @staticmethod
    #define a function
    def encode(string_input):
        #function for string input validation
        def string_validation(string_input):
            #check the constraints of the input code:
            #1. if the string is empty, return empty string
            if not string_input:
                return ""

            #2. string_input length must be between 0-50 characters:
            if not 0 <= len(string_input) <= 50:
                raise ValueError("String input length must be between 0 and 50, no more.")

            #3. string_input can only consists of letters 'a-z' and 'A-Z' but can also accept digits, special characters, and white space:
            ch_allowed = set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 {}[]():;'+=.,")
            if not all(char in ch_allowed for char in string_input):
                raise ValueError("There are invalid characters in the string_input")
        #check input
        string_validation(string_input)

        def character_encode(string_input):
            #creating a list to store ch
            ch_list = []

            #count of num of consecutive characters
            num_consecutive_ch = 1

            #check if string input is empty:
            if not string_input:
                return ""

            #loop
            for i in range(1, len(string_input)):
                #if current ch is equal to the before
                if string_input[i] == string_input[i-1]:
                    #increment by 1
                    num_consecutive_ch +=1

                else:
                    #num of consecutive character is greater than 4
                    if num_consecutive_ch > 4:
                        #add the format to list
                        ch_list.append(f"/{num_consecutive_ch:02d}{string_input[i-1]}")

                    else: #if num of consecutive character is 4 or less
                        ch_list.append(string_input[i-1] * num_consecutive_ch)
                    num_consecutive_ch = 1
            #last ch of string input
            if num_consecutive_ch > 4:
                ch_list.append(f"/{num_consecutive_ch:02d}{string_input[-1]}")
            else:
                ch_list.append(string_input[-1] * num_consecutive_ch)

            return ''.join(ch_list)

        return character_encode(string_input)


#asking user for string input:
def user_in():
    return input("Please enter a string to encode: ")

#code the input to the class
def code_string():
    in_userstring = user_in()
    codestring = RunLengthEncode.encode(in_userstring)
    return codestring

#print result
def print_string(codestring):
    print("The encoded string is: " , codestring)

codestring = code_string()
print_string(codestring)

File: RunLengthEncode.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 */

package com.mycompany.runlengthencode;

/**
 *
 * @author eliza
 */
import javax.swing.JOptionPane;

/**
 *
 * @author eliza
 */
public class RunLengthEncode
{
    public static String encode(String input) throws IllegalArgumentException
    {
        //string validation:
        string_validation(input);
         
        return codeCH(input);
    }
    
    private static void string_validation(String input) throws IllegalArgumentException
    {
        //constraints check:
        
        //1. if the string is empty, return empty
        if(input.isEmpty())
        {
            return;
        }
        
        //2. check if length of string is between 0 <= 50:
        if(!(input.length() >= 0 && input.length() <= 50))
        {
            throw new IllegalArgumentException("The length of the string"
                    + " must be between 0 and 50 characters please:)");
        }
        
        //3. 'a-z' and 'A-Z' allowed, but can accept special characters and digits:
        if(!input.matches("[a-zA-Z0-9 {}\\[\\]():;'+=.,]*"))
        {
            throw new IllegalArgumentException("Please input valid characters. Thank you.");
        }
    }
    
    public static String codeCH(String input)
    {
        //sequence of characters
        StringBuilder chList = new StringBuilder();
        //num of consecutive characters
        int num_consecutive_ch = 1;
                
        //looping:
        for(int i = 1; i <= input.length(); i++)
        {
            //if current ch == before ch
            if(i < input.length() && input.charAt(i) == input.charAt(i-1))
            {
                num_consecutive_ch++; //increment
            }
            else
            {
                //if # of consecutive ch is > 4:
                if(num_consecutive_ch > 4)
                {
                    chList.append(String.format("/%02d%c", num_consecutive_ch, input.charAt(i-1)));
                }
                else
                {
                    //# of consecutive ch is 4 or less
                    for (int j = 0; j < num_consecutive_ch; j++)
                    {
                        chList.append(input.charAt(i-1));
                    }
                }
                //reset
                num_consecutive_ch = 1;
            }
        }
        
        
        return chList.toString();
    }
    
    
    
    public static void main(String[] args)
    {
        String input = UserInput();
        String codestring = encode(input);
        Print(codestring);
        
    }
    
    //JDialog Box for user input
    private static String UserInput()
    {
        return JOptionPane.showInputDialog(null, "Please enter "
                + "a string to encode: ");
        
    }
    
    //print function:
    private static void Print(String codestring)
    {
        JOptionPane.showMessageDialog(null, "The encoded "
                + "string is: " + codestring);
    }
}

